### ISSUE TITLE
Use <webview> to handle multiple operations on a single Nightmare (or electron) instance?

### ISSUE THREAD

**Issue by danielnieto** [segmentio/nightmare#828]
I was wondering if it's possible to use Electron's <webView>s to load the webpages and adding/removing those tags dynamically inside a single BrowserWindow to handle async requests, (like if each webview was a single instance). I just came across Nightmare while I was researching a way to scrap from Electron, the thing is that I want to scrap lots of links (+100 links, yeah, like data-mining), my first approach was to use <webView>'s (with injected "preload" scripts and returning HTML code through IPC messages ) and it was good, but since in the future I might use authentication, handle sessions clicking on stuff and waiting for dynamic loads, and more complicated flows I looked for more "sophisticated" ways to do it, and I found out about PhantomJS (I used it with Horseman) It was a total disappointment, my raw webview "scrapper" took about ~2 seconds to return the whole HTML from a webpage, while PhantomJS took more than ~10 seconds, then I found Nightmare and use it and it takes about ~5 seconds to accomplish the same... all good if you're using it with a single page, but if I try to load those 100+ links (by creating a new instance in a loop, I know that Nightmare do not support more than 1 request at the same time on a single instance) it would take too much time (compared to the <webview> approach), and I believe that most of the time spent is consumed on launching a new instance, and not loading the actual webpage... So I thought that it could be possible to have only 1 instance of Nightmare (Electron) but somehow have several <webview>'s on it and manage them independently? Have you guys thought about it? Is there technical limitations on this approach? Is this something you plan to support in the future?

**Comment by rosshinkley** [comment]
Short answer is yes, it's definitely on the list of things to try, and I'd point you at #593 for more information.

Longer answer: it's very possible to manage multiple `BrowserWindow`s from a single Electron instance, and should be reasonably straightforward to do with Nightmare.  It would solve situations where you're time-bound (like yours, as you noted, launching new Chromium instances takes quite a bit of time) or memory-bound (each Chromium instance is _very_ expensive).  It would also solve problems with managing popup windows and testing with multiple windows.

The downside is that (I think) it would be a breaking change to Nightmare 2.x, which I'm not against, but it would have to be handled ... carefully.

**Comment by danielnieto** [comment]
oh, great!... in the meantime I created a project precisely for these kind "scrapping" needs:

[Scrapman](https://github.com/danielnieto/scrapman)

it is based on Electron as well, but it uses the technique described above, btw, I couldn't ever imagine all the painful process needed to communicate Node -> electron(main) -> electron(render) -> shadow DOM with only async messages! I had to create another [NPM module](https://github.com/danielnieto/ipc-messages-manager) just to handle the IPC messaging.

I really really hope that these features get implemented on Nightmare in the future! Looking forward to it!
